stl/inc/format

```diff
@@ -343,72 +337,398 @@ public:
+        return static_cast<long long>(42);
+    } else if constexpr (unsigned_integral<_Ty> && sizeof(_Ty) <= sizeof(unsigned long long)) {
+        return static_cast<unsigned long long>(42);
+    } else if constexpr (is_same_v<_Ty, bool>) {
+        return true;
+    } else if constexpr (is_same_v<_Ty, _CharType>) {
+        return _CharType{};
+    } else if constexpr (is_same_v<_Ty, float>) {
+        return float{};
+    } else if constexpr (is_same_v<_Ty, double>) {
+        return double{};
+    } else if constexpr (is_same_v<_Ty, long double>) {
+        return static_cast<long double>(42);
+    } else if constexpr (is_same_v<_Ty, const void*>) {
+        return static_cast<const void*>(nullptr);
+    } else if constexpr (is_same_v<_Ty, const _CharType*>) {
+        return static_cast<const _CharType*>(nullptr);
+    } else if constexpr (is_same_v<_Ty, basic_string_view<_CharType>>) {
+        return basic_string_view<_CharType>{};
+    } else {
+        return basic_format_arg<_Context>::handle();
+    }
+}
+
+template <class _Context, class _Ty>
+static constexpr size_t _Get_format_arg_storage_size() noexcept {
+    return sizeof(_Get_format_arg_storage_type<_Context, _Ty>());
+}
+
+struct _Format_arg_store_packed_index {
+    // TRANSITION, Should be templated on number of arguments for even less storage
+    using _Index_type = size_t;
+
+    constexpr _Format_arg_store_packed_index() = default;
+    constexpr _Format_arg_store_packed_index(const size_t _Index)
+        : _Index(static_cast<_Index_type>(_Index)), _Type(_Basic_format_arg_type::_None) {}
+
+    _Index_type _Index : (sizeof(_Index_type) * 8 - 4);
+    _Basic_format_arg_type _Type : 4;
 };
 
-// TODO: test coverage
 template <class _Context, class... _Args>
 class _Format_arg_store {
-    static constexpr size_t _Num_args = sizeof...(_Args);
-    // TODO: no arg packing yet
+private:
+    template <class>
+    friend class basic_format_args;
 
-    using _Value_type = basic_format_arg<_Context>;
+    static constexpr size_t _Num_args       = sizeof...(_Args);
+    static constexpr size_t _Storage_length = (_Get_format_arg_storage_size<_Context, _Args>() + ... + 0);
+
+    using _CharType   = typename _Context::char_type;
+    using _Index_type = _Format_arg_store_packed_index;
+
+    template <class _Ty, _Basic_format_arg_type _Arg_type>
+    void _Store_impl(const size_t _Arg_index, _Ty _Val) noexcept {
+        const auto _Store_index = _Index_array[_Arg_index]._Index;
+        const auto _Length      = _Get_format_arg_storage_size<_Context, _Ty>();
+
+        std::memcpy(_Storage + _Store_index, _STD addressof(_Val), _Length);
+        _Index_array[_Arg_index]._Type = _Arg_type;
+        if (_Arg_index < _Num_args) {
+            _Index_array[_Arg_index + 1] = _Format_arg_store_packed_index{_Store_index + _Length};
+        }
+    }
+
+    byte _Storage[(_STD max)(_Storage_length, 1ull)];
+    _Index_type _Index_array[(_STD max)(_Num_args, 1ull)];

```

Comment:
Hm that might be true

Reply:
We need to store it in memory, as `basic_format_args::get` doesn't have access to enough type information to reconstruct the storage, as it is type erased

Reply:
So i was thinking about this when I totally should have went to sleep and to no ones surprise @CaseyCarter is absolutely right, that we can get around it. Kind off though...

1. We cannot get fully rid of it because `basic_format_args` holds a type erased pointer to `format-arg-store`. We do not have the type information available anymore. That means we need at least the `_Basic_format_arg_type` portion of the `_Index_type`, otherwise `basic_format_args::get` is in trouble

2. I screwed up and ordered the members. `_Index_array` is static in the sense that `_Index_type` has fixed size, so we can simply flip it to the front and do not really need to hold both pointers in `basic_format_args`, as the second array starts directly after the first. To be save of any strange padding issues I would still put them into a single `byte` array though.

3. Tradeoffs. Currently we have random access to each `basic_format_arg` through the `_Index` portion of `_Index_type`. We could theoreticall get rid of that and simply store the `_Basic_format_arg_type` of each `basic_format_arg` in the front and then do a linear probe to get the offset of the stored value at position `_Idx`.

This would allow us to get supe compact storage at a runtime cost. I "guess" that the number of format arguments will be small, but I think this will be unacceptably slow for larger numbers.

So for now I would go with `_Index_array` + `_Storage` in a single `byte` array and optimize that to use a smaller base type for `_Index_type`